### PR TITLE
mimalloc: add version 1.7.7/2.0.7, support conan v2

### PR DESCRIPTION
**Pull request by toge**
Specify library name and version:  **mimalloc/***

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/mimalloc/all/conandata.yml [hunk 1]
```diff
@@ -21,28 +27,23 @@ sources:
...
   "2.0.3":
     - patch_file: "patches/2.0.3-0001-change-install-paths-avoid-symlink.patch"
-      base_path: "source_subfolder"
   "2.0.2":
     - patch_file: "patches/2.0.2-0001-change-install-paths-avoid-symlink.patch"
-      base_path: "source_subfolder"
     - patch_file: "patches/2.0.2-0002-include-cstddef-to-get-std-size-t.patch"
-      base_path: "source_subfolder"
+  "1.7.7":
+    - patch_file: "patches/1.7.7-0001-change-install-paths-avoid-symlink.patch"
```
Comment (prince-chrismc) [inline recipes/mimalloc/all/conandata.yml:42]: Patch fields please!

File: recipes/mimalloc/all/conanfile.py [hunk 2]
```diff
@@ -56,56 +53,54 @@ def _compilers_minimum_version(self):
...
-            if self.options.get_safe("inject"):
-                del self.options.inject
+            self.options.rm_safe("single_object")
+            self.options.rm_safe("inject")
+
+    def layout(self):
+        cmake_layout(self, src_folder="src")
 
     def validate(self):
         # Currently, mimalloc/1.7.6,2.0.6 does not work properly with shared MD builds.
```
Comment (prince-chrismc) [inline recipes/mimalloc/all/conanfile.py:90]:
```suggestion
        # Currently, mimalloc some version do not work properly with shared MD builds.
```

File: recipes/mimalloc/all/conanfile.py [hunk 3]
```diff
@@ -56,56 +53,54 @@ def _compilers_minimum_version(self):
...
-        if  self.version in ["1.7.6", "2.0.6"] and \
+        if  self.version in ["1.7.6", "1.7.7", "2.0.6", "2.0.7"] and \
             self.options.shared and \
-            microsoft.is_msvc(self) and \
-            "MD" in microsoft.msvc_runtime_flag(self):
+            is_msvc(self) and \
+            "MD" in msvc_runtime_flag(self):
             raise ConanInvalidConfiguration(
-                "Currently, mimalloc/1.7.6,2.0.6 doesn't work properly with shared MD builds.")
+                f"Currently, {self.ref} doesn't work properly with shared MD builds.")
```
Comment (prince-chrismc) [inline recipes/mimalloc/all/conanfile.py:97]:
```suggestion
                f"Currently, {self.ref} doesn't work properly with shared MD builds in CCI. Contributions welcomed")
```

File: recipes/mimalloc/all/conanfile.py [hunk 4]
```diff
@@ -114,81 +109,78 @@ def validate(self):
...
+            rm(self, "*.a", os.path.join(self.package_folder, "lib"))
             shutil.move(os.path.join(self.package_folder, self._obj_name + ".o"),
                         os.path.join(self.package_folder, "lib"))
             shutil.copy(os.path.join(self.package_folder, "lib", self._obj_name + ".o"),
                         os.path.join(self.package_folder, "lib", self._obj_name))
 
         if self.settings.os == "Windows" and self.options.shared:
             if self.settings.arch == "x86_64":
-                self.copy("mimalloc-redirect.dll", src=os.path.join(self._source_subfolder, "bin"),
+                self.copy("mimalloc-redirect.dll", src=os.path.join(self.source_folder, "bin"),
```
Comment (prince-chrismc) [inline recipes/mimalloc/all/conanfile.py:169]:
```suggestion
                copy(self, "mimalloc-redirect.dll", src=os.path.join(self.source_folder, "bin"),
```

File: recipes/mimalloc/all/conanfile.py [hunk 5]
```diff
@@ -114,81 +109,78 @@ def validate(self):
...
                           dst="bin")
             elif self.settings.arch == "x86":
-                self.copy("mimalloc-redirect32.dll", src=os.path.join(self._source_subfolder, "bin"),
+                self.copy("mimalloc-redirect32.dll", src=os.path.join(self.source_folder, "bin"),
```
Comment (prince-chrismc) [inline recipes/mimalloc/all/conanfile.py:172]:
```suggestion
                copy(self, "mimalloc-redirect32.dll", src=os.path.join(self.source_folder, "bin"),
```

File: recipes/mimalloc/all/test_package/conanfile.py [hunk 6]
```diff
@@ -1,44 +1,86 @@
-from conans import ConanFile, CMake, RunEnvironment, tools
-import os
+from conan import ConanFile
+from conan.tools.build import can_run
+from conan.tools.cmake import cmake_layout, CMake
```
Comment (prince-chrismc) [inline recipes/mimalloc/all/test_package/conanfile.py:3]: Duplicate of below

```suggestion
```

**Review by prince-chrismc — COMMENTED** [review]
Just some minor nits but it looks really good

File: recipes/mimalloc/all/conandata.yml (hunk 1 above)
Comment (toge) [reply recipes/mimalloc/all/conandata.yml]: @prince-chrismc 
Thank you for your review!
I added patch_descriptions.

**Review by uilianries — COMMENTED** [review]
- please, use `        if self.options.override and \` with self.info.options.override at line 107

**Review by prince-chrismc — DISMISSED** [review]

**Review by prince-chrismc — APPROVED** [review]

**Review by uilianries — APPROVED** [review]
LGTM

**Pull request merged**